File: python/785. Is Graph Bipartite?.py
class Solution(object):
    def isBipartite(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: bool
        """
        n = len(graph)
        colors = [0]*n

        def dfs(idx, color):
            if colors[idx]!=0:
                return colors[idx]==color
            colors[idx] = color
            for i in graph[idx]:
                if not dfs(i, -color):
                    return False
            return True

        for i in range(n):
            if colors[i]==0 and not dfs(i, 1):
                return False
        return True


if __name__ == '__main__':
    graph = [[1,3],[0,2],[1,3],[0,2]]
    su = Solution()
    su.isBipartite(graph)

File: go/src/leetcode/785/785. Is Graph Bipartite?.go
/*
*
@author ZhengHao Lou
Date    2022/12/05
*/
package main

/*
*
https://leetcode.cn/problems/is-graph-bipartite/
*/
func isBipartite(graph [][]int) bool {
	n := len(graph)
	colors := make([]int, n)
	var dfs func(idx, color int) bool
	dfs = func(idx, color int) bool {
		if colors[idx] != 0 {
			return colors[idx] == color
		}
		colors[idx] = color
		for _, ni := range graph[idx] {
			if !dfs(ni, -color) {
				return false
			}
		}
		return true
	}

	for i := 0; i < n; i++ {
		if colors[i] == 0 && !dfs(i, 1) {
			return false
		}
	}

	return true
}
